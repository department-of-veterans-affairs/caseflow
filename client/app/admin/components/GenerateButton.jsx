/* eslint-disable func-style */
import React, { useEffect, useState } from 'react';
import Button from '../../components/Button';
import Modal from '../../components/Modal';
import Alert from '../../components/Alert';
import LoadingContainer from '../../components/LoadingContainer';
import { LOGO_COLORS } from 'app/constants/AppConstants';
import PropTypes from 'prop-types';
import { CSVLink } from 'react-csv';

const GenerateButton = (props) => {

  // state properties
  const [showConfirmModal, setShowConfirmModal] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [showEmptyResultsBanner, setShowEmptyResultsBanner] = useState(false);
  const [showErrorBanner, setShowErrorBanner] = useState(false);
  const [extractedResults, setExtractedResults] = useState('');

  useEffect(() => {
    setIsLoading(props.isLoading);
  }, [props.isLoading]);

  useEffect(() => {
    setExtractedResults(props.extractedResults);

    if (props.extractedResults?.length > 0) {
      setShowEmptyResultsBanner(false);
      setShowConfirmModal(true);
    }
  }, [props.extractedResults]);

  useEffect(() => {
    if (props.emptyResultsMessage?.length > 0) {
      setShowEmptyResultsBanner(true);
      setShowConfirmModal(false);
    }
  }, [props.emptyResultsMessage]);

  useEffect(() => {
    if (props.manualExtractionSuccess === false) {
      setShowErrorBanner(true);
    } else {
      setShowErrorBanner(false);
    }
  }, [props.manualExtractionSuccess]);

  const onClickGenerate = () => {
    props.sendExtractRequest();
  };

  return (
    <div style={{ height: '75vh' }}>
      {
        showEmptyResultsBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="No Veterans were found" type="success" />
        </div>
      }
      {
        showErrorBanner &&
        <div style={{ padding: '10px' }}>
          <Alert message="Veteran Extract Failed" type="error" />
        </div>
      }
      {
        !isLoading &&
        <Button
          id="generate-extract"
          onClick={() => onClickGenerate()}
        >
          Generate Veteran Extract
        </Button>
      }
      {
        showConfirmModal &&

        <Modal title="This file contains PII"
          confirmButton={<CSVLink data={extractedResults} filename="veteran_extract.csv" onClick={() => setShowConfirmModal(false)}>Confirm</CSVLink>}
          closeHandler={() => {
            setShowConfirmModal(false);
          }}
        >
          Are you sure you want to download it?
        </Modal>
      }
      {
        isLoading &&

        <LoadingContainer color={LOGO_COLORS.QUEUE.ACCENT}>
          <div className="loading-div">
            Action is Running...
          </div>
        </LoadingContainer>
      }
    </div>
  );
};

GenerateButton.propTypes = {
  extractedResults: PropTypes.string,
  manualExtractionSuccess: PropTypes.bool,
  isLoading: PropTypes.bool,
  sendExtractRequest: PropTypes.func,
  emptyResultsMessage: PropTypes.string,
};

export default GenerateButton;
