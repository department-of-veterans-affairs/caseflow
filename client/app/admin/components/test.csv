VET_LAST_NAME,VET_FIRST_NAME,VET_MIDDLE_NAME,VET_DATE_OF_BIRTH,VET_SSN,VET_PARTICIPANT_ID,VET_FILE_NUMBER,APPELLANT_LAST_NAME,APPELLANT_FIRST_NAME,APPELLANT_MIDDLE_NAME,APPELLANT_DATE_OF_BIRTH,APPELLANT_SSN,APPELLANT_GENDER,APPELLANT_ADDRESS,APPELLANT_PHONE,APPELLANT_EMAIL,APPELLANT_EDI_PI,APPELLANT_CORP_PID,APPELLANT_VACOLS_INTERNAL_ID,RELATIONSHIP_TO_VETERAN