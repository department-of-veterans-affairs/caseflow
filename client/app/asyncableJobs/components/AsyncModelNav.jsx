import React from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import DropdownButton from '../../components/DropdownButton';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

export default class AsyncModelNav extends React.PureComponent {
  modelNameLinks = () => {
    const links = [];
    const modelNames = this.props.models.sort();
    const numLinks = modelNames.length;

    for (let modelName of modelNames) {
      const url = `/asyncable_jobs/${modelName}/jobs`;
      let modelLink;

      if (numLinks > 4) {
        modelLink = {
          title: modelName,
          target: url
        };
      } else {
        modelLink = <span key={modelName} className="cf-model-jobs-link"><a href={url}>{modelName}</a></span>;
      }

      links.push(modelLink);
    }

    return numLinks > 4 ? <DropdownButton lists={links} label="Filter by Job Type" /> : links;
  }

  render = () => {
    const currentPath = this.props.asyncableJobKlass ?
      `/asyncable_jobs/${this.props.asyncableJobKlass}/jobs` :
      '/jobs';

    return <div>
      <strong>Last updated:</strong> {moment(this.props.fetchedAt).format(DATE_TIME_FORMAT)}
      <div style={{ marginTop: '.5em' }}>
        <a style={{ marginRight: '.5em' }} href="/jobs" className="cf-link-btn">All jobs</a>
        {this.modelNameLinks()}
        <a style={{ float: 'right' }} href={`${currentPath}.csv`} className="cf-link-btn">Download as CSV</a>
      </div>
    </div>;
  }
}

AsyncModelNav.propTypes = {
  models: PropTypes.array,
  fetchedAt: PropTypes.string,
  asyncableJobKlass: PropTypes.string
};
