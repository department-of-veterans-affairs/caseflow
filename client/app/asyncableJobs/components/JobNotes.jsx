import React from 'react';
import Button from '../../components/Button';
import ApiUtil from '../../util/ApiUtil';
import PropTypes from 'prop-types';
import moment from 'moment';
import ReactMarkdown from 'react-markdown';

const DATE_TIME_FORMAT = 'ddd MMM DD HH:mm:ss YYYY';

import Checkbox from '../../components/Checkbox';
import TextareaField from '../../components/TextareaField';

class NewNoteForm extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      saveInProgress: false,
      checkboxSelected: true,
      newNote: ''
    };
  }

  onSave = () => {
    const job = this.props.job;
    const url = `/asyncable_jobs/${job.klass}/jobs/${job.id}/note`;
    const form = this;
    const notes = this.props.notes;

    const data = {
      note: this.state.newNote,
      send_to_intake_user: this.state.checkboxSelected
    };

    notes.setState({ reloaded: false });

    ApiUtil.post(url, { data }).
      then(
        (response) => {
          const newNote = JSON.parse(response.text);

          form.setState({
            saveInProgress: false,
            checkboxSelected: false,
            newNote: ''
          });

          // add note to current page.
          notes.props.notes.unshift(newNote);
          notes.setState({ reloaded: true });
        }
      );
  }

  onChange = (event) => {
    this.setState({ newNote: event });
  }

  onCheckboxChange = (event) => {
    this.setState({ checkboxSelected: event });
  }

  isSaveDisabled = () => {
    return this.state.saveInProgress || this.state.newNote.trim().length === 0;
  }

  isCheckboxChecked = () => {
    return this.state.checkboxSelected;
  }

  sendToUserCheckbox = () => {
    return <Checkbox
      label="Send as message to user"
      name="send_to_intake_user"
      value={this.isCheckboxChecked()}
      onChange={this.onCheckboxChange}
      disabled={this.state.saveInProgress}
    />;
  }

  render = () => {
    return <div className="comment-size-container">
      <TextareaField
        hideLabel
        name="Add Note"
        aria-label="Add Note"
        onChange={this.onChange}
        value={this.state.newNote}
      />
      <div className="comment-save-button-container">
        <span className="cf-right-side">
          { this.props.job.user && this.sendToUserCheckbox() }
          <Button
            name="save"
            disabled={this.isSaveDisabled()}
            onClick={this.onSave}>
              Add Note
          </Button>
        </span>
      </div>
    </div>;
  }

}

NewNoteForm.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.object,
  onSave: PropTypes.func,
  onCancel: PropTypes.func
};

export default class JobNotes extends React.PureComponent {

  render = () => {
    const { notes, job } = this.props;

    return <div className="job-notes">
      <h3>Notes</h3>
      <NewNoteForm job={job} notes={this} />
      <div>
        { notes.map((note, index) => {
          return <div className="job-note" key={`job-note-container-${index}`}>
            <div
              className="job-note-details"
              data-key={`job-note-${index}`}
              key={`job-note-${index}`}
              id={`job-note-${note.id}`}>

              <div>
                <span className="job-note-time">{moment(note.created_at).format(DATE_TIME_FORMAT)}</span>
                <span className="job-note-user">{note.user}</span>
              </div>
              <div className="job-note-note"><ReactMarkdown source={note.note} /></div>

            </div>
          </div>;
        })}
      </div>
    </div>;
  }
}

JobNotes.propTypes = {
  job: PropTypes.object,
  notes: PropTypes.array
};
