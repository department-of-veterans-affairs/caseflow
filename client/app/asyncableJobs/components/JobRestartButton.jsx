import React from 'react';
import Button from '../../components/Button';
import ApiUtil from '../../util/ApiUtil';
import PropTypes from 'prop-types';

class JobRestartButton extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      restarted: false,
      restarting: false
    };
  }

  restartJob = () => {
    this.setState({ restarting: true });
    this.sendRestart();
  }

  sendRestart = () => {
    const button = this;
    const page = this.props.page;
    const job = this.props.job;
    const url = `/asyncable_jobs/${job.klass}/jobs/${job.id}`;

    ApiUtil.patch(url, {}).
      then(
        (response) => {
          const responseObject = JSON.parse(response.text);

          Object.assign(job, responseObject);

          job.error = '';
          job.restarted = true;

          button.setState({
            restarted: true,
            restarting: false
          });

          // trigger table rerender
          const jobsRestarted = page.state.restarted + 1;

          page.setState({ restarted: jobsRestarted });
        }
      );
  }

  getButtonClassNames = () => {
    let classNames = ['usa-button'];

    if (this.state.restarting || this.state.restarted) {
      classNames.push('usa-button-disabled');
    }

    return classNames;
  }

  getButtonText = () => {
    let txt = 'Restart';

    if (this.state.restarting) {
      txt = 'Restarting';
    } else if (this.state.restarted) {
      txt = 'Restarted';
    } else if (this.props.job.processed_at) {
      txt = 'Processed';
    } else if (this.disableRestart()) {
      txt = 'Queued';
    }

    return txt;
  }

  disableRestart = () => {
    const job = this.props.job;

    if (job.processed_at) {
      return true;
    }

    if (!job.attempted_at) {
      return true;
    }

    const fiveMinutes = 300000;

    const lastAttempted = new Date(job.attempted_at).getTime();
    const submittedAt = new Date(job.last_submitted_at).getTime();
    const now = new Date().getTime();

    if ((now - lastAttempted) < fiveMinutes || (now - submittedAt) < fiveMinutes) {
      return true;
    }

    return false;
  }

  render = () => {
    const job = this.props.job;

    return <Button
      id={`job-${job.klass}-${job.id}`}
      title={`${job.klass} ${job.id}`}
      loading={this.state.restarting}
      loadingText="Restarting..."
      disabled={this.disableRestart()}
      onClick={() => {
        this.restartJob();
      }}
      classNames={this.getButtonClassNames()}
    >{this.getButtonText()}</Button>;
  }
}

JobRestartButton.propTypes = {
  job: PropTypes.object,
  page: PropTypes.object
};

export default JobRestartButton;
