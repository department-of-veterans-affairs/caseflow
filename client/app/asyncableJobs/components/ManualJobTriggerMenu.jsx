import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Button from '../../components/Button';
import StringUtil from '../../util/StringUtil';
import classNames from 'classnames';
import LoadingContainer from '../../components/LoadingContainer';
import { LOGO_COLORS } from '../../constants/AppConstants';

const ManualJobTriggerMenu = (props) => {

  const [loading, setLoading] = useState(false);
  const [manualJobType, setManualJobType] = useState(null);

  useEffect(() => {
    setManualJobType(props.manualJobType);
    if (props.manualJobType) {
      setLoading(false);
    }

  }, [props.manualJobType]);

  const renderMessage = () => {
    if (manualJobType) {
      const manualJobBannerClasses = classNames({
        success: props.manualJobSuccess,
        fail: !props.manualJobSuccess,
      }, 'manual-job-banner');

      const status = props.manualJobSuccess ? 'succeeded' : 'failed';

      return (
        <div className={manualJobBannerClasses}>
          Manual run of <strong>{StringUtil.snakeCaseToCapitalized(manualJobType)}</strong> has{' '}
          <strong>{status}</strong>
        </div>
      );
    }

    return null;
  };

  return (
    <div className="tab-border scheduled-jobs-tab">
      <br />
      <h1>Manually Perform Scheduled Async Jobs</h1>
      {loading &&
        <LoadingContainer color={LOGO_COLORS.QUEUE.ACCENT}>
          <div className="loading-div">
            Job is Running...
          </div>
        </LoadingContainer>
      }
      {!loading && (<>
        {renderMessage()}
        <table>
          <tbody className="manual-job-table">
            {props.supportedJobs.map((jobType) => (
              <tr key={`${jobType}-row`}>
                <td>
                  <h3>{StringUtil.snakeCaseToCapitalized(jobType)}</h3>
                </td>
                <td>
                  <Button
                    name={`trigger-${jobType}-job`}
                    onClick={() => {
                      setLoading(true);
                      props.sendJobRequest(jobType);
                    }}
                  >
                Perform Now
                  </Button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </>)}
    </div>
  );
};

ManualJobTriggerMenu.propTypes = {
  supportedJobs: PropTypes.arrayOf(
    PropTypes.string
  ),
  sendJobRequest: PropTypes.func.isRequired,
  manualJobStatus: PropTypes.number,
  manualJobSuccess: PropTypes.bool,
  manualJobType: PropTypes.string,
};

export default ManualJobTriggerMenu;
