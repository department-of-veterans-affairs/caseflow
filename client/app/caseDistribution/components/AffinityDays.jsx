import React, { useState, useEffect } from 'react';
import { useSelector } from 'react-redux';
import cx from 'classnames';
import NumberField from 'app/components/NumberField';
import TextField from 'app/components/TextField';
import COPY from '../../../COPY';
import ACD_LEVERS from '../../../constants/ACD_LEVERS';
import { getUserIsAcdAdmin } from '../reducers/levers/leversSelector';

const AffinityDays = () => {
  const theState = useSelector((state) => state);
  const isUserAcdAdmin = getUserIsAcdAdmin(theState);

  const storeLevers = useSelector((state) => state.caseDistributionLevers.levers.affinity);
  const [affinityLevers, setAffinityLevers] = useState(storeLevers);

  useEffect(() => {
    setAffinityLevers(storeLevers);
  }, [storeLevers]);

  const generateFields = (dataType, option, lever) => {
    const useAriaLabel = !lever.is_disabled_in_ui;
    const tabIndex = lever.is_disabled_in_ui ? -1 : 0;

    if (dataType === ACD_LEVERS.data_types.number) {
      return (
        <NumberField
          name={option.item}
          title={option.text}
          label={option.unit}
          isInteger
          readOnly={lever.is_disabled_in_ui ? true : (lever.value !== option.item)}
          value={option.value}
          errorMessage={option.errorMessage}
          onChange={() => console.warn('not implemented')}
          id={`${lever.item}-${option.value}`}
          inputID={`${lever.item}-${option.value}-input`}
          useAriaLabel={useAriaLabel}
          tabIndex={tabIndex}
          disabled={lever.is_disabled_in_ui}
        />
      );
    }
    if (dataType === ACD_LEVERS.data_types.text) {
      return (
        <TextField
          name={option.item}
          title={option.text}
          label={false}
          readOnly={lever.is_disabled_in_ui ? true : (lever.value !== option.item)}
          value={option.value}
          onChange={() => console.warn('not implemented')}
          id={`${lever.item}-${option.value}`}
          inputID={`${lever.item}-${option.value}-input`}
          useAriaLabel={useAriaLabel}
          tabIndex={tabIndex}
          disabled={lever.is_disabled_in_ui}
        />
      );
    }

    return null;
  };

  const generateMemberViewLabel = (option, lever, index) => {
    if (lever.value === option.item) {
      return (
        <div key={`${option.item}-${lever.item}-${index}`}>
          <div>
            <label id={lever.item}
              className={lever.is_disabled_in_ui ? 'lever-disabled' : 'lever-active'}
              htmlFor={`${lever.item}-${option.item}`}
            >
              {`${option.text} ${option.data_type === ACD_LEVERS.data_types.number ?
                `${option.value} ${option.unit}` : ''}`}
            </label>
          </div>
        </div>
      );
    }

    return null;
  };

  const renderAdminInput = (option, lever, index) => {
    const className = cx('combined-radio-input', (lever.value === option.item) ? '' : 'outline-radio-input');

    return (
      <div key={`${lever.item}-${index}-${option.item}`}>
        <div>
          <input
            checked={option.item === lever.value}
            type={ACD_LEVERS.data_types.radio}
            value={option.item}
            disabled={lever.is_disabled_in_ui}
            id={`${lever.item}-${option.item}`}
            name={lever.item}
            onChange={() => console.warn('not implemented')}
          />
          <label htmlFor={`${lever.item}-${option.item}`}>
            {option.text}
          </label>
        </div>
        <div>
          <div className={className} aria-label={option.unit}>
            {generateFields(option.data_type, option, lever, isUserAcdAdmin)}
          </div>
        </div>
      </div>
    );
  };

  affinityLevers?.sort((leverA, leverB) => leverA.lever_group_order - leverB.lever_group_order);

  return (
    <div className="lever-content">
      <div className="lever-head">
        <h2>{COPY.CASE_DISTRIBUTION_AFFINITY_DAYS_H2_TITLE}</h2>
        <div className="lever-left">
          <strong>{COPY.CASE_DISTRIBUTION_BATCH_SIZE_LEVER_LEFT_TITLE}</strong>
        </div>
        <div className="lever-right">
          <strong>{COPY.CASE_DISTRIBUTION_BATCH_SIZE_LEVER_RIGHT_TITLE}</strong>
        </div>
      </div>
      {affinityLevers.map((lever, index) => (
        <div className={cx('active-lever', lever.is_disabled_in_ui ? 'lever-disabled' : '')}
          key={`${lever.item}-${index}`}
        >
          <div className="lever-left">
            <strong>{lever.title}</strong>
            <p className="affinity-lever-text">{lever.description}</p>
          </div>
          <div id={lever.item}
            className={cx('lever-right', 'affinity-lever-num-sec')}
          >
            {lever.options.map((option) => (
              (isUserAcdAdmin) ? renderAdminInput(option, lever, index) : generateMemberViewLabel(option, lever, index)
            ))}
          </div>
        </div>
      ))}
      <h4 className="footer-styling">{COPY.CASE_DISTRIBUTION_FOOTER_ASTERISK_DESCRIPTION}</h4>
      <div className="cf-help-divider"></div>
    </div>
  );
};

export default AffinityDays;
