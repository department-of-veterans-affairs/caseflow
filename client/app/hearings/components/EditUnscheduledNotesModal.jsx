import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { sprintf } from 'sprintf-js';

import COPY from '../../../COPY';
import { onReceiveAmaTasks } from '../../queue/QueueActions';
import { showErrorMessage, showSuccessMessage } from '../../queue/uiReducer/uiActions';

import ApiUtil from '../../util/ApiUtil';
import Modal from '../../components/Modal';
import Button from '../../components/Button';
import { UnscheduledNotes } from './UnscheduledNotes';

export const EditUnscheduledNotesModal = ({
  task,
  appeal,
  onCancel,
  ...props
}) => {
  const [loading, setLoading] = useState(false);
  const [unscheduledNotes, setUnscheduledNotes] = useState(task?.unscheduledHearingNotes?.notes);

  const disable = loading || unscheduledNotes === task?.unscheduledHearingNotes?.notes;

  const submit = async () => {
    try {
      const data = {
        task: {
          business_payloads: {
            values: {
              notes: unscheduledNotes
            }
          }
        }
      };

      setLoading(true);
      // Add the google analytics event
      window.analyticsEvent('Hearings', 'Add/edit notes', 'Case Details');

      await ApiUtil.patch(`/tasks/${task.taskId}`, { data }).then((resp) => {
        props.onReceiveAmaTasks(resp.body.tasks.data);
        props.showSuccessMessage({
          title: sprintf(COPY.SAVE_UNSCHEDULED_NOTES_SUCCESS_MESSAGE, appeal?.veteranFullName),
          detail: null
        });
      });
    } catch (err) {
      const error = {
        title: COPY.SAVE_UNSCHEDULED_NOTES_ERROR_TITLE,
        detail: COPY.SAVE_UNSCHEDULED_NOTES_ERROR_DETAIL
      };

      props.showErrorMessage(error);
    } finally {
      setLoading(false);
      // close the modal
      onCancel();
      // Focus the top of the page to display alert
      window.scrollTo(0, 0);
    }
  };

  return (
    <Modal
      title="Edit Notes"
      closeHandler={onCancel}
      confirmButton={<Button disabled={disable} onClick={submit}>Save</Button>}
      cancelButton={<Button linkStyling disabled={loading} onClick={onCancel}>Cancel</Button>}
    >
      <UnscheduledNotes
        onChange={(notes) => setUnscheduledNotes(notes)}
        unscheduledNotes={unscheduledNotes}
        updatedAt={task?.unscheduledHearingNotes?.updatedAt}
        updatedByCssId={task?.unscheduledHearingNotes?.updatedByCssId}
        uniqueId={task?.taskId}
      />
    </Modal>
  );
};

EditUnscheduledNotesModal.propTypes = {
  task: PropTypes.object,
  appeal: PropTypes.object,
  onCancel: PropTypes.func,
  onReceiveAmaTasks: PropTypes.func,
  showErrorMessage: PropTypes.func,
  showSuccessMessage: PropTypes.func
};

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      onReceiveAmaTasks,
      showErrorMessage,
      showSuccessMessage
    },
    dispatch
  );

export default connect(null, mapDispatchToProps)(EditUnscheduledNotesModal);
