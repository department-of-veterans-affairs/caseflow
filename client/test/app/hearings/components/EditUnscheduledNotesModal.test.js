import React from 'react';
import { render, screen } from '@testing-library/react';
import { axe } from 'jest-axe';

import { generateAmaTask } from 'test/data/tasks';
import { amaAppeal } from 'test/data/appeals';

import { EditUnscheduledNotesModal } from 'app/hearings/components/EditUnscheduledNotesModal';

describe('EditUnscheduledNotesModal', () => {
  const onCancel = jest.fn();
  const hearingTask = generateAmaTask({
    uniqueId: '3',
    type: 'HearingTask',
    status: 'on_hold'
  })
  const defaultProps = {
    task: hearingTask,
    appeal: amaAppeal,
    onCancel: onCancel
  }

  it('renders correctly', () => {
    const { container } = render(<EditUnscheduledNotesModal {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it('passes a11y testing', async () => {
    const { container } = render(<EditUnscheduledNotesModal {...defaultProps} />);

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('displays pre filled data', () => {
    const maxCharLimit = 1000
    const notes = "Pre filled notes"
    const hearingTaskWithNotesData = {
      ...hearingTask,
      unscheduledHearingNotes: {
        updatedAt: '2020-09-08T10:03:49.210-04:00',
        updatedByCssId: 'BVASYELLOW',
        notes: notes
      }
    }
    render(
      <EditUnscheduledNotesModal {...defaultProps} task={hearingTaskWithNotesData}/>
    );

    expect(screen.getByLabelText('Notes')).toBeInTheDocument()
    expect(screen.getByText(notes)).toBeInTheDocument()
    const charLimitMessage = `${maxCharLimit - notes.length} characters left`
    expect(screen.getByText(charLimitMessage)).toBeInTheDocument()
    expect(screen.getByText('Last updated by BVASYELLOW on 09/08/2020'))
      .toBeInTheDocument()
  })
})
