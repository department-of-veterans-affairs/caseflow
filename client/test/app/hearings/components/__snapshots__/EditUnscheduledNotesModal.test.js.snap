// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditUnscheduledNotesModal renders correctly 1`] = `
<div>
  <section
    aria-describedby="modal_id-desc"
    aria-labelledby="modal_id-title"
    aria-modal="true"
    class="cf-modal active "
    id="modal_id"
    role="dialog"
  >
    <div
      class="cf-modal-body"
      id=""
    >
      <button
        class="cf-modal-close"
        id="Edit-Notes-button-id-close"
        type="button"
      >
        <span
          class="usa-sr-only"
        >
          Close
        </span>
        <svg
          class="cf-icon-close"
          height="55"
          viewBox="0 0 55 55"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title>
            close
          </title>
          <g
            fill="#212121"
          >
            <path
              d="M52.6 46.9l-6 6c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-13-13-13 13c-.8.8-1.9 1.2-3 1.2s-2.2-.4-3-1.2l-6-6c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l13-13-13-13c-.8-.8-1.2-1.9-1.2-3s.4-2.2 1.2-3l6-6c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l13 13 13-13c.8-.8 1.9-1.2 3-1.2s2.2.4 3 1.2l6 6c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3l-13 13 13 13c.8.8 1.2 1.9 1.2 3s-.4 2.2-1.2 3z"
            />
          </g>
        </svg>
      </button>
      <div
        style="display: flex;"
      >
        <div
          data-css-1gs0ko2=""
        >
          <h1
            id="modal_id-title"
          >
            Edit Notes
          </h1>
          <div
            data-css-o98ubo=""
          >
            <div>
              <div
                class="cf-form-textarea"
                data-css-1tc952m=""
              >
                <label
                  class="question-label"
                  data-css-42cjjw=""
                  for="null-unscheduled-notes"
                >
                  <strong>
                    <span>
                      Notes
                    </span>
                  </strong>
                </label>
                <textarea
                  id="null-unscheduled-notes"
                  maxlength="1000"
                  name="null-unscheduled-notes"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="cf-modal-divider"
      />
      <div
        class="cf-modal-controls"
      >
        <div>
          <span
            class="cf-push-right"
          >
            <span>
              <button
                class="cf-submit usa-button-disabled usa-button"
                disabled=""
                type="button"
              >
                Save
              </button>
            </span>
          </span>
          <span
            class="cf-push-left"
          >
            <span>
              <button
                class="cf-submit cf-btn-link usa-button"
                type="button"
              >
                Cancel
              </button>
            </span>
          </span>
        </div>
      </div>
    </div>
  </section>
</div>
`;
